user/man/evlsend.1
.\" This manpage has been modified by Dennis Grace
.\" of IBM LTC Documentation Services from a document
.\" that was automatically generated by docbook2man
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/>
.\" Please send any bug reports, improvements, comments, patches,
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.\" For questions regarding the modifications from the
.\" docbook2man output, contact Dennis Grace at
.\" <dgrace@us.ibm.com>.
.TH "EVLSEND" "" "13 February 2003" "" ""
.SH NAME
evlsend \- event generation utility
.SH SYNOPSIS
\fBevlsend\fR \fB -f | --facility \fIfacility\fB \fR
\fB -t | --type  \fIevent_type\fB
\fR[\fB -s | --severity \fIseverity\fB \fR]
.br
[\fB -m | --message \fImessage_string\fB \fR]
.br
[\fB -p | --printf \fIformat_string\fB \fIattr_value\fB \fI ...\fR]
.br
[\fB -b | --binary \fIattr_type\fB \fIattr_value\fB \fR\fI ...\fR]

.SH "DESCRIPTION"
.PP
The \fBevlsend\fR utility lets you send an
event to the event-logging system.
Specify \fB--message\fP, \fB--printf\fP, or \fB--binary\fP
to indicate the format and contents of the event record.
.SH "OPTIONS"
.TP
\fB-f, --facility \fIfacility\fB\fR
Facility to record in the event
record. This option must be specified, and the
\fIfacility\fR must be specified
in the Facility Registry.
.TP
\fB-t, --type \fIevent_type\fB\fR
Specify the event type to record in
the event record. This option must be specified, except that it is optional
when \fB--printf\fP is specified.
.TP
\fB-s, --severity \fIseverity\fB\fR
Severity to record in the event
record. If this option is not specified, then the
default severity is INFO. Valid severity values are
EMERG, ALERT, CRIT, ERR, WARNING, NOTICE, INFO,
and DEBUG.
.TP
\fB-m, --message \fImessage_string\fB \fR
Specifies that the variable data
format is POSIX_LOG_STRING and logs the message.
.TP
\fB-p, --printf \fIformat_string\fB \fIattr_value ...\fR
Specifies that the variable data
format is POSIX_LOG_PRINTF.
Logs the format string and zero or more attribute values.
The number and type of attribute values must match the format string.
If the \fB--type\fP option is omitted, the record's event type
is set to a checksum of the format string.
.TP
\fB-b, --binary \fIattr_type\fB \fIattr_value ...\fR
Specifies that the variable data
format is POSIX_LOG_BINARY.
\fIattr_type\fR and
\fIattr_value\fR must both be
specified, and multiple pairs of these two parameters
can be specified.
Each type/value pair specifies a value to be packed
(without padding for alignment)
into the variable portion
of the event record.
The list of type/value pairs may optionally be terminated with an
.I attr_type
of endofdata, which is discarded.
Any
.I attr_type
that starts with a dash (-) terminates the list
and is interpreted as a command-line option.

The
.B --binary
option supports the following data types:
char, uchar,
short, ushort,
int, uint,
long, ulong,
longlong, ulonglong,
address,
float, double, ldouble,
string,
wchar, wstring.

Integral values (including char, uchar, wchar, and address)
can be expressed in decimal, hexadecimal, or octal.
For a value of type char, uchar, or wchar, if
.I attr_value
is not a valid number but is a single character,
the ASCII value of that character is used.
If
.I attr_type
is wstring, the string
.I attr_value
is converted to a wide-character string.

For
.I n
consecutive values of the type
.I typename
(e.g., an array),
.I attr_type
can be specified as
\fIn\fR*\fItypename\fR
or as
.IR typename []
followed by
.IR n .
Both notations are equivalent.
The following
.I attr_value
must be
.I n
values of the indicated type.
.SH "EXAMPLES"
.TP
\fBLogging a string:\fR
.nf
evlsend -f LOCAL3 -s INFO -t 61 -m "Log this string"
                
.fi
.TP
\fBLogging binary data:\fR
.nf
evlsend --facility LOCAL5 --type 65 --binary uchar 0x22 string "Log
this string too, but use default severity."
                
evlsend -f local5 -t 66 -b 'string[]' 3 decimal hex octal '3*int' 25 0x19 031
.fi
.TP
\fBLogging printf-format data:\fR
.nf
evlsend -f user -p 'Command %s failed with status=%d' $cmd $status
.fi
.SH "FILES"
.TP
\fB\fI/var/evlog/eventlog\fB\fR
Standard Event log
.TP
\fB\fI/var/evlog/privatelog\fB\fR
Private Event log


CHANGES
Enterprise-Class Event Logging for Linux
========================================

v1.0.0   9/20/2001
Initial  Version


v1.0.1  10/01/2001
- Replaced calls to spin_lock_irq() and spin_unlock_irq() with calls
  to  spin_lock_irqsave() and spin_unlock_irqrestore() in kernel log
  write functions.  This resolved a hang condition experienced with 
  a SCSI device driver.
- Replaced use of sockaddr to sockaddr_un to provide longer socket 
  path name.
- Several minor changes needed for s390 (32 and 64 bit).
- runtests.sh scripts modifed to take option of s390 (32-bit) or
  s390x (64-bit) to specify s390/IBM Z-Series.
- Processing of white-spaces was fixed to correctly process >1 white 
  space.
NOTE:  The patch for GNU C Library, evlog-1.0.1_glibc2.2.3.patch,
is unchanged from version 1.0.0.


v1.1.0  11/09/2001

New features and enhancements in v1.1.0
    
1. Kernel patches for kernel versions 2.4.2 through 2.4.14 will 
   now be posted on the website.  
2. A new library that "wraps" around GNU C Lib and provides
   forwarding of syslog messages to the POSIX event log without
   having to patch and rebuild glibc.
3. Formatting Template support and evltc command (for compiling
   formatting template). Some features are not yet available
   (section numbers in the spec are indicated):
* (6.1.2.2.1.2: Imports and Typedefs) If an import statement specifies
   a struct that has already been created or imported, the import
   statement will be silently ignored.
* (6.1.2.2.1.4: Attributes Section [overall format]) The issue regarding
   const-only structs is still unresolved.  It is currently impossible 
   to define a struct that has only const attributes.
* (6.1.2.2.1.4.3: Attribute Dimension or Bit-Field Width) Bit fields 
   and arrays of structs are not yet supported.
* (6.1.2.2.1.4.5 Attribute Format)   The following is not yet supported:
* Validation of printf-style formats is incomplete.  In particular, you
  are permitted to specify formats that are completely inappropriate for
  the given data type -- e.g. "%f" for a string.
* The %t (dump) format is currently supported only for arrays.
* %b (bitmap) and %v (named-value) specifiers are not yet supported
  for arrays.
* As previously mentioned, arrays of structs are not yet supported;
  so the %Z format specifier is not, either.
* (6.2.3.3 Formatting Templates [API]) The following is not supported:
   * The evl_populatetemplate() function does not currently create the 
     _EXTRA_DATA_ pseudo-attribute
   * The evltemplate_getatt() function does not currently handle names 
     such as "x.y" to reference members of struct-type attributes.

4. Displaying of events logged in the kernel (with the new log write
   functions) on the system console.  The --output option was added to 
   the evlconfig command to turn console logging on/off, or set the
   severity level for displaying to the console.
5. Forwarding of printk() messages to the kernel event buffer must now
   be explicitly enabled when configuring the kernel (forwarding is
   disabled by default).
6. evlfacility command for managing the facility registry.
7. Restricted Logging support. 
8. Added "address" as a data type for formatting templates.
9. Added 2 new purpose arguments for posix_log_query_create() as
   extensions to the draft POSIX standard:  EVL_PRPS_TEMPLATE and
   EVL_PRPS_RESTRICTED.
10. Added EVL_KERNEL_EVENT flag for log_flags member in 
    posix_log_entry.
11. Added varargs versions of log write functions in kernel and 
    user-space.
12. Added support for --new and --timeout options in evlview
    command.
13. Added --templates and --notemplates options to evlview command.
14. Added unlink(PidFile) function in evlogd.c, evlnotifyd.c,
    evlactiond.c.
15. Moved evlogd, evlnotifyd, and evlactiond pid files from /var/evlog
    to /var/run.

Bug Fixes in v1.1.0

1. Removed evlog *.pid files on graceful shutdown or on startup if 
   previous shutdown was non-graceful to fix evlogd daemon startup
   failure.
2. In event notification, fixed a bug where error returns from
   rt_sigqueueinfo() were causing the notification to be erroneously
   disabled.
3. In do_syslog(), now use options 20 and 21 for read/write to kernel
   event buffer, instead of 9 and 10, to avoid conflict with other
   newly added options in do_syslog(). 
4. posix_log_notify_add() is now signal-safe.


v1.1.1  11/28/2001

Bug fixes in v1.1.1

1. (Fix for Bug # 485676) Code added in v1.1.0 for displaying events on the
   system console is incompatible with the re-write of code for displaying of
   messages from printk() first appearing in kernel v2.4.10. This causes the
   kernel to hang or PANIC under certain conditions.  This only affects the
   patches for kernel versions 2.4.10 through 2.4.14. 
2. Added check in the kernel version of posix_log_write() for null-terminated
   string (when log_format is POSIX_LOG_STRING).  Returns EBADMSG if string
   is not null-terminated.  A few other changes to make this function   
   consistent between kernel and user space.
3. Fixed bug causing programs linked with libevl.so to fail at link time.
4. Added missing function call in libevlsyslog.so so that syslog messages
   actually get logged in event log.
5. Fixed several bugs causing problems on s390 and/or 64-bit machines:
   * evltc command reporting bogus errors or causing segfaults
   * Backward seeks through the event log (as used by evlview --tail) failed. 
   * Various memory corruption problems 
6. Updated all header files for C++ compatability.


v1.2.0  02/18/2002

New Features and Enhancements in v1.2.0

1. Events logged in user-space are now passed directly to the evlogd 
   daemon,  instead of being written to the kernel event buffer.
2. The libevl event logging library no longer depends on the event 
   logging kernel patch, meaning that event logging in user-space 	
   does not require  changes in the kernel.  The kernel patch is 

user/cmd/evlsend/evlsend.c
/*
 * Linux Event Logging for the Enterprise
 * Copyright (C) International Business Machines Corp., 2002
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include "posix_evlog.h"
#include "posix_evlsup.h"

extern int _evlPackCmdArgs(int argc, const char **argv, int *iarg,
	char *databuf, size_t *datasz, char *errbuf, size_t ebufsz);
extern int _evlPackCmdArgsPerFormat(const char *format, int argc,
	const char **argv, int *iarg, char *databuf, size_t *datasz,
	char *errbuf, size_t ebufsz);

const char *progName = 0;

static void
usage()
{
	fprintf(stderr,
"Usage: %s -f | --facility facility -t | --type event_type\n"
"             [-s | --severity severity]\n"
"             [-m | --message message]\n"
"             [-p | --printf format attr-value ...]\n"
"             [-b | --binary attr-type attr-value ...]\n",
		progName);
	exit(1);
}

/*
 * If s is an integer (and nothing but an integer), return its value.
 * Else return badVal.  If unSigned == 1, require an unsigned integer.
 */
static long
extractNumber(const char *s, long badVal, int unSigned)
{
	char *endPtr = 0;
	long nl;
	if (unSigned) {
		nl = strtoul(s, &endPtr, 0);
	} else {
		nl = strtol(s, &endPtr, 0);
	}

	if (*endPtr != '\0') {
		return badVal;
	}
	return nl;
}

/*
 * arg is either a name or a sequence of digits.  If it's a name in the
 * indicated table, return the corresponding value.  Else try to
 * decode it as a number.  If that fails, return badVal.  If it's a valid
 * number, verify that it's in the table; return badVal if it isn't.
 */
static int
getNumberOrName(const char *arg, const struct _evlNvPair table[], int badVal,
	int unSigned)
{
	int n;

	if (table) {
		n = _evlGetValueByCIName(table, arg, badVal);
		if (n != badVal) {
			return n;
		}
	}

	/* Not a valid name in the indicated table.  Should be a number. */
	n = (int) extractNumber(arg, badVal, unSigned);
	if (n != badVal) {
		/* It's a number.  Is it in the table? */
		const struct _evlNvPair *nv;
		for (nv = table; nv->nv_name; nv++) {
			if (nv->nv_value == n) {
				return n;
			}
		}
	}
	return badVal;
}

main(int argc, char **argv)
{
	int i, status;
	posix_log_facility_t facility = EVL_INVALID_FACILITY;
	int type = EVL_INVALID_EVENT_TYPE;
	posix_log_severity_t severity = LOG_INFO;
	const char *msg = NULL;
	const char *format = NULL;
	int haveFac = 0;
	int haveEvt = 0;
	int haveSev = 0;
	int haveMsg = 0;
	int haveBinary = 0;
	int havePrintf = 0;
	char databuf[POSIX_LOG_ENTRY_MAXLEN];
	size_t reclen = 0;
	struct _evlNvPair *facilities = _evlSnapshotFacilities();

	progName = argv[0];

	for (i = 1; i < argc; i++) {
		if (!strcmp(argv[i], "-f") || !strcmp(argv[i], "--facility")) {
			if (++i >= argc || haveFac++) {
				usage();
			}
			facility = getNumberOrName(argv[i], facilities,
				EVL_INVALID_FACILITY, 1);
			if (facility == EVL_INVALID_FACILITY) {
fprintf(stderr, "%s: unknown facility: %s\n", progName, argv[i]);
				exit(1);
			}
		} else if (!strcmp(argv[i], "-t") || !strcmp(argv[i], "--type")) {
			if (++i >= argc || haveEvt++) {
				usage();
			}
			/* Event type must be numeric. */
			type = extractNumber(argv[i], EVL_INVALID_EVENT_TYPE, 0);
			if (type == EVL_INVALID_EVENT_TYPE) {
fprintf(stderr, "%s: bad event type: %s\n", progName, argv[i]);
				exit(1);
			}
		} else if (!strcmp(argv[i], "-s") || !strcmp(argv[i], "--severity")) {
			if (++i >= argc || haveSev++) {
				usage();
			}
			/* Severity can be numeric or symbolic. */
			severity = getNumberOrName(argv[i], _evlSeverities, -1, 1);
			if (severity == -1) {
fprintf(stderr, "%s: unknown severity: %s\n", progName, argv[i]);
				exit(1);
			}
		} else if (!strcmp(argv[i], "-m") || !strcmp(argv[i], "--message")) {
			if (++haveMsg + haveBinary + havePrintf != 1) {
				usage();
			}
			if (++i > argc) {
				usage();
			}
			msg = argv[i];
		} else if (!strcmp(argv[i], "-b") || !strcmp(argv[i], "--binary")) {
			char errbuf[100];
			if (haveMsg + ++haveBinary + havePrintf != 1) {
				usage();
			}
			i++;
			if (_evlPackCmdArgs(argc, (const char**)argv, &i,
			    databuf, &reclen, errbuf, 100) != 0) {
				fprintf(stderr, "%s: %s\n", progName, errbuf);
				exit(1);
			}
			if (i < argc) {
				/*
				 * There are more args to process.  i now
				 * points to the next arg after the -b list.
				 * i-- for now because the for statement
				 * will promptly i++.
				 */
				i--;
			}
		} else if (!strcmp(argv[i], "-p") || !strcmp(argv[i], "--printf")) {
			char errbuf[100];
			if (haveMsg + haveBinary + ++havePrintf != 1) {
				usage();
			}
			if (++i >= argc) {
				usage();
			}
			format = argv[i++];
			if (_evlPackCmdArgsPerFormat(format, argc,
			    (const char**)argv, &i, databuf, &reclen,
			    errbuf, 100) != 0) {
				fprintf(stderr, "%s: %s\n", progName, errbuf);
				exit(1);
			}
			if (i < argc) {
				/* See above comment. */
				i--;
			}
		} else {
			usage();
		}
	}

	if (!haveFac) {
		usage();
	}

	if (!haveEvt) {
		if (havePrintf) {
			type = evl_gen_event_type_v2(format);
		} else {
			usage();
		}
	}

	if (haveMsg) {
		status = posix_log_write(facility, type, severity, msg,
			strlen(msg)+1, POSIX_LOG_STRING, 0);
	} else if (haveBinary) {
		status = posix_log_write(facility, type, severity, databuf,
			reclen, (reclen ? POSIX_LOG_BINARY : POSIX_LOG_NODATA), 
			0);
	} else if (havePrintf) {
		status = posix_log_write(facility, type, severity, databuf,
			reclen, POSIX_LOG_PRINTF, 0);
	} else {
		usage();
	}
	if (status != 0) {
		errno = status;
		perror("posix_log_write");
		exit(1);
	}
	exit(0);
}
